jwst/dark_current/tests/test_dark_sub.py

```diff
@@ -0,0 +1,517 @@
+RAPID['nskip'] = 0
+
+READPATTERNS = {}
+READPATTERNS['DEEP8'] = DEEP8
+READPATTERNS['DEEP2'] = DEEP2
+READPATTERNS['MEDIUM8'] = MEDIUM8
+READPATTERNS['MEDIUM2'] = MEDIUM2
+READPATTERNS['SHALLOW4'] = SHALLOW4
+READPATTERNS['SHALLOW2'] = SHALLOW2
+READPATTERNS['BRIGHT2'] = BRIGHT2
+READPATTERNS['BRIGHT1'] = BRIGHT1
+READPATTERNS['RAPID'] = RAPID
+
+TFRAME = 10.73677
+
+
+def test_frame_averaging(setup_nrc_cube):
+    '''Check that if nframes>1 or groupgap>0, then the pipeline reconstructs
+       the dark reference file to match the frame averaging and groupgap
+       settings of the exposure.'''
+
+    # Values to build the fake data arrays
+    ngroups = 5
+    nrows = 2048
+    ncols = 2048
+
+    # Loop over the NIRCam readout patterns:
+    for name in READPATTERNS:
+
+        # Get the configuration for the readout pattern
+        readpatt = name
+        nframes = READPATTERNS[readpatt]['nframes']
+        groupgap = READPATTERNS[readpatt]['nskip']
+
+        # Create data and dark model
+        data, dark = setup_nrc_cube(readpatt, ngroups, nrows, ncols)
+
+        # Add ramp values to dark model data array
+        dark.data[:, 500, 500] = np.arange(0, 100)
+        dark.err[:, 500, 500] = np.arange(100, 200)
+
+        # Run the pipeline's averaging function
+        avg_dark = average_dark_frames(dark, ngroups, nframes, groupgap)
+
+        # Group input groups into collections of frames which will be averaged
+        total_frames = (nframes * ngroups) + (groupgap * (ngroups-1))
+
+        # Get starting/ending indexes of the input groups to be averaged
+        gstrt_ind = np.arange(0, total_frames, nframes + groupgap)
+        gend_ind = gstrt_ind + nframes
+
+        # Prepare arrays to hold results of averaging
+        manual_avg = np.zeros((ngroups))
+        manual_errs = np.zeros((ngroups))
+
+        # Manually average the input data to compare with pipeline output
+        for newgp, gstart, gend in zip(range(ngroups), gstrt_ind, gend_ind):
+
+            # Average the data frames
+            newframe = np.mean(dark.data[gstart:gend, 500, 500])
+            manual_avg[newgp] = newframe
+
+            # ERR arrays will be quadratic sum of error values
+            manual_errs[newgp] = np.sqrt(np.sum(dark.err[gstart:gend, 500, 500]**2)) / (gend - gstart)
+
+        # Check that pipeline output matches manual averaging results
+        assert np.all(manual_avg == avg_dark.data[:, 500, 500])
+        assert np.all(manual_errs == avg_dark.err[:, 500, 500])
+
+        # Check that meta data was properly updated
+        assert avg_dark.meta.exposure.nframes == nframes
+        assert avg_dark.meta.exposure.ngroups == ngroups
+        assert avg_dark.meta.exposure.groupgap == groupgap
+
+
+
+def test_more_sci_frames():
+    '''Check that data is unchanged if there are more frames in the science data is than in the reference file'''
+
+    # size of integration
+    nints = 1
+    ngroups = 30
+    xsize = 1032
+    ysize = 1024
```

Comment:
I have a question about this. The dark reference files are subarray specific. If we start just putting in random array sizes, won't this cause a problem? Are you saying we should be doing all these tests on subarrays instead?

Reply:
It's true that if you were using actual dark reference files in these tests you'd need to worry about whether they're subarray or full-frame. But since all of the tests create both the science data and reference data arrays on the fly, all that matters is that the dimensions match one another (at least the 2 spatial dimensions). The low-level routines that do the actual data manipulation have no knowledge of (and don't care about) whether the data came from full-frame or subarray.

Reply:
Ok, you're right. I forgot that we set these up to create the dark data models as well.